File: opcua-webservice-node/src/server/service/transactions/reactiveobject/UaBrowseMemberTransaction.ts
import { NodeClass, StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaBrowseDescription, UaExpandedNodeId, UaReferenceDescription, UaNodeId } from "opcua-webapi-ts";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaReactiveObjectType } from "../../../addressspace/reactiveobject/UaReactiveObjectType";
import {
    UaBrowseAdditionalInfo,
    UaChildIdentifier,
    UaInstanceIdentifier,
    UaObjectId,
    UaObjectIdentifier,
} from "../../../types";
import { BrowseMemberRequest, BrowseMemberResponse } from "../../message";
import { UaBrowseTransaction } from "../base/UaBrowseTransaction";
import { ServiceContext } from "../../..";

export class UaBrowseMemberTransaction extends UaBrowseTransaction {
    private readonly objectType: UaReactiveObjectType;
    private readonly objectId: UaObjectId;
    private readonly memberId: UaChildIdentifier;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        serviceContext: ServiceContext,
        nodeToBrowse: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
        handleId: number,
        objectType: UaReactiveObjectType,
        objectId: UaObjectId,
        memberId: UaChildIdentifier,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this.objectType = objectType;
        this.objectId = objectId;
        this.memberId = memberId;
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            if (!this._additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) || this.memberId.pathL2 !== null) {
                return;
            }

            const isMethod = this.memberId.methodNode;
            const request = new BrowseMemberRequest(
                this.objectId,
                this.memberId.path,
                isMethod,
                this.getItem(),
            );

            const response = await this.objectType.onBrowseMemberChildren(request);
            this.browseMemberChildResult(isMethod, response);
        } catch (error) {
            this.buildErrorResults(error);
        }
    }

    private browseMemberChildResult(isMethodParent: boolean, response: BrowseMemberResponse): void {
        for (const item of response.children) {
            if (item.id.length === 0 || !item.isForward || item.nodeClass !== NodeClass.Variable) {
                continue;
            }

            const objectIdentifier = new UaObjectIdentifier(
                this.objectType.nodeId.toString(),
                this.objectId.id,
                this.objectId.instance?.nodeId.toString() ?? null,
            );

            const memberIdentifier = new UaChildIdentifier(
                this.memberId.path,
                item.id,
                isMethodParent,
            );

            const newIdentifier = new UaInstanceIdentifier(objectIdentifier, memberIdentifier);
            const byteString = newIdentifier.toByteString();
            if (byteString === null) {
                this._statusCode = makeUaStatusCode(StatusCodes.BadNodeIdInvalid);
                continue;
            }

            this._references.push(
                new UaReferenceDescription(
                    UaExpandedNodeId.from(new UaNodeId(byteString, this.nodeManager.nsIndex())),
                    NodeClass.Variable,
                    item.browseName,
                    item.displayName,
                    item.referenceTypeId,
                    item.isForward,
                    UaExpandedNodeId.from(item.typeDefinitionId),
                ),
            );
        }
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/reactiveobject/UaBrowseMemberTransaction.java
package org.opcfoundation.webserver.service.transactions.reactiveobject;

import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.QualifiedName;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;
import org.eclipse.milo.opcua.stack.core.types.structured.ReferenceDescription;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.service.transactions.base.UaBrowseTransaction;
import org.opcfoundation.webapi.service.types.ServiceContext;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseMemberRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseMemberResponse;
import org.opcfoundation.webserver.types.common.*;

import java.util.concurrent.CompletableFuture;

public class UaBrowseMemberTransaction extends UaBrowseTransaction {
    private final UaReactiveObjectType objectType;
    private final UaObjectId objectId;
    private final UaChildIdentifier memberId;
    private final NodeManagerReactiveObject nodeManager;

    public UaBrowseMemberTransaction(
            ServiceContext serviceContext,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId,
            UaReactiveObjectType objectType,
            UaObjectId objectId,
            UaChildIdentifier memberId,
            NodeManagerReactiveObject nodeManager)
    {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this.objectType = objectType;
        this.objectId = objectId;
        this.memberId = memberId;
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        try {
            if (!additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) ||
                    null != memberId.getPathL2()) return CompletableFuture.completedFuture(null);

            boolean isMethod = (null != memberId.getMethodNode()) && memberId.getMethodNode();

            BrowseMemberRequest request = new BrowseMemberRequest(
                    objectId,
                    memberId.getPath(),
                    isMethod,
                    getItem());

            return objectType.onBrowseMemberChildren(request).
                    thenApply(response -> { return browseMemberChildResult(isMethod, response); }).
                    exceptionally(ex -> buildErrorResponse(ex.getCause()));
        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private Void browseMemberChildResult(boolean isMethodParent, BrowseMemberResponse response)
    {
        for (UaReferenceDescriptor item: response.getChildren())
        {
            if (item.getId().isEmpty() || !item.isForward() || item.getNodeClass() != NodeClass.Variable) continue;

            UaObjectIdentifier objectIdentifier = new UaObjectIdentifier(
                        objectType.nodeId().toParseableString(),
                        objectId.getId(),
                        (null == objectId.getInstance()) ? null : objectId.getInstance().nodeId().toParseableString());

            UaChildIdentifier memberIdentifier = new UaChildIdentifier(
                    memberId.getPath(),
                    item.getId(),
                    isMethodParent);

            UaInstanceIdentifier newIdentifier = new UaInstanceIdentifier(
                        objectIdentifier,
                        memberIdentifier);

            // System.out.println("Result: " + newIdentifier);

            references.add(new ReferenceDescription(
                    item.getReferenceTypeId(),
                    item.isForward(),
                    new NodeId(nodeManager.nsIndex(), newIdentifier.toByteString()).expanded(),
                    new QualifiedName(0,item.getBrowseName()),
                    item.getDisplayName(),
                    NodeClass.Variable,
                    item.getTypeDefinitionId().expanded()));
        }

        return null;
    }
}
